proxy/proxy.go

```diff
@@ -401,11 +401,14 @@ func (p *Proxy) Resolve(d *DNSContext) error {
 	// Use cache only if it's enabled and the query doesn't use custom
 	// upstreams.
 	cacheWorks := p.cache != nil && d.CustomUpstreamConfig == nil
-	if cacheWorks {
-		if p.replyFromCache(d) {
+	if cacheWorks && !d.disableCacheReading {
+		hit, isExpired := p.replyFromCache(d)
+		if hit && (p.Config.CacheOptimistic || !isExpired) {
+			if isExpired {
+				go p.resolveAndUpdateCache(d)
```

Comment:
In fact, it will only schedule a call when optimistic caching is enabled and the cached data has expired. At this time, it will directly return the cached data. So it will not affect the number of calls to upstream servers.

Or do you mean that  the expired domain  should be resolved only once before  the cache is updated?

Reply:
The resolution will take some time (let's say 100-200ms). Until it happens every new DNS query to this domain will trigger a new goroutine (which is not really needed since you've already scheduled it once).

Reply:
I thought about doing this before, just a bit lazy...